File: php/src/Lox/Scanner.php
<?php

namespace Nkoll\Plox\Lox;

use IntlChar;
use Nkoll\Plox\PloxCommand;

class Scanner
{
    /** @var Token[] */
    private array $tokens = [];
    private string $source;

    private int $start = 0;
    private int $current = 0;
    private int $line = 1;

    private const array keywords = [
        "and" => TokenType::AND,
        "class" => TokenType::class,
        "else" => TokenType::ELSE,
        "false" => TokenType::FALSE,
        "for" => TokenType::FOR,
        "fun" => TokenType::FUN,
        "if" => TokenType::IF,
        "nil" => TokenType::NIL,
        "or" => TokenType::OR,
        "print" => TokenType::PRINT,
        "return" => TokenType::RETURN,
        "super" => TokenType::SUPER,
        "this" => TokenType::THIS,
        "true" => TokenType::TRUE,
        "var" => TokenType::VAR,
        "while" => TokenType::WHILE,
    ];

    public function __construct(string $source)
    {
        $this->source = $source;
    }

    /** @return Token[]  */
    public function scanTokens(): array
    {
        while(!$this->isAtEnd()) {
            $this->start = $this->current;
            $this->scanToken();
        }

        $this->tokens[] = new Token(TokenType::EOF, "", null, $this->line);
        return $this->tokens;
    }

    private function isAtEnd(): bool
    {
        return $this->current >= strlen($this->source);
    }

    private function scanToken(): void
    {
        $c = $this->advance();
        switch($c) {
            case '(':
                $this->addToken(TokenType::LEFT_PAREN);
                break;
            case ')':
                $this->addToken(TokenType::RIGHT_PAREN);
                break;
            case '{':
                $this->addToken(TokenType::LEFT_BRACE);
                break;
            case '}':
                $this->addToken(TokenType::RIGHT_BRACE);
                break;
            case ',':
                $this->addToken(TokenType::COMMA);
                break;
            case '.':
                $this->addToken(TokenType::DOT);
                break;
            case '-':
                $this->addToken(TokenType::MINUS);
                break;
            case '+':
                $this->addToken(TokenType::PLUS);
                break;
            case ';':
                $this->addToken(TokenType::SEMICOLON);
                break;
            case '*':
                $this->addToken(TokenType::STAR);
                break;
            case '!':
                $this->addToken($this->match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
                break;
            case '=':
                $this->addToken($this->match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
                break;
            case '<':
                $this->addToken($this->match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
                break;
            case '>':
                $this->addToken($this->match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
                break;
            case '/':
                if ($this->match('/')) {
                    while($this->peek() !== "\n" && !$this->isAtEnd()) {
                        $this->advance();
                    }
                    break;
                }
                $this->addToken(TokenType::SLASH);
                break;
            case ' ':
            case "\r":
            case "\t":
                // Ignore whitespace.
                break;
            case "\n":
                $this->line++;
                break;
            case '"':
                $this->string();
                break;
            default:
                if (IntlChar::isdigit($c)) {
                    $this->number();
                    break;
                } elseif (IntlChar::isalpha($c)) {
                    $this->identifier();
                    break;
                }
                PloxCommand::error($this->line, "Unexpected Character.");
        }
    }

    private function string(): void
    {
        while($this->peek() !== '"' && !$this->isAtEnd()) {
            if ($this->peek() === "\n") {
                $this->line++;
            }
            $this->advance();
        }

        if ($this->isAtEnd()) {
            PloxCommand::error($this->line, "Unterminated String.");
            return;
        }

        $this->advance();

        $value = $this->substr($this->source, $this->start + 1, $this->current - 1);
        $this->addToken(TokenType::STRING, $value);
    }

    private function number(): void
    {
        while(IntlChar::isdigit($this->peek())) {
            $this->advance();
        }

        if ($this->peek() === '.' && IntlChar::isdigit($this->peekNext())) {
            $this->advance();

            while(IntlChar::isdigit($this->peek())) {
                $this->advance();
            }
        }

        $value = floatval($this->substr($this->source, $this->start, $this->current));
        $this->addToken(TokenType::NUMBER, $value);
    }

    private function identifier(): void
    {
        while(IntlChar::isalnum($this->peek())) {
            $this->advance();
        }

        $value = $this->substr($this->source, $this->start, $this->current);
        $type = self::keywords[$value] ?? TokenType::IDENTIFIER;
        $this->addToken($type);
    }

    private function match(string $expected): bool
    {
        if ($this->isAtEnd()) {
            return false;
        }
        if ($this->source[$this->current] !== $expected) {
            return false;
        }

        $this->current++;
        return true;
    }

    private function peek(): string
    {
        if ($this->isAtEnd()) {
            return "\0";
        }
        return $this->source[$this->current];
    }

    private function peekNext(): string
    {
        if ($this->current + 1 >= strlen($this->source)) {
            return "\0";
        }
        return $this->source[$this->current + 1];
    }


    private function advance(): string
    {
        return $this->source[$this->current++];
    }

    private function addToken(TokenType $type, $literal = null): void
    {
        $text = $this->substr($this->source, $this->start, $this->current);
        $this->tokens[] = new Token($type, $text, $literal, $this->line);
    }

    private function substr(string $source, int $begin, int $end): string
    {
        return substr($source, $begin, $end - $begin);
    }

}

File: zlox/src/lox/Scanner.zig
const std = @import("std");

const Scanner = @This();

start: usize,
current: usize,
source: []const u8,
line: usize,

pub fn init(source: []const u8) Scanner {
    return .{
        .start = 0,
        .current = 0,
        .source = source,
        .line = 1,
    };
}

pub fn scanToken(self: *Scanner) Token {
    self.start = self.current;

    if (self.isAtEnd()) {
        return self.makeToken(.EOF);
    }

    const c = self.advanceWithValue();
    if (isDigit(c)) {
        return self.number();
    }

    if (isAlpha(c)) {
        return self.identifier();
    }

    switch (c) {
        '(' => return self.makeToken(.LEFT_PAREN),
        ')' => return self.makeToken(.RIGHT_PAREN),
        '{' => return self.makeToken(.LEFT_BRACE),
        '}' => return self.makeToken(.RIGHT_BRACE),
        ';' => return self.makeToken(.SEMICOLON),
        ',' => return self.makeToken(.COMMA),
        '.' => return self.makeToken(.DOT),
        '-' => return self.makeToken(.MINUS),
        '+' => return self.makeToken(.PLUS),
        '/' => return self.makeToken(.SLASH),
        '*' => return self.makeToken(.STAR),
        '!' => {
            if (self.match('=')) {
                return self.makeToken(.BANG_EQUAL);
            }
            return self.makeToken(.BANG);
        },
        '=' => {
            if (self.match('=')) {
                return self.makeToken(.EQUAL_EQUAL);
            }
            return self.makeToken(.EQUAL);
        },
        '<' => {
            if (self.match('=')) {
                return self.makeToken(.LESS_EQUAL);
            }
            return self.makeToken(.LESS);
        },
        '>' => {
            if (self.match('=')) {
                return self.makeToken(.GREATER_EQUAL);
            }
            return self.makeToken(.GREATER);
        },
        '"' => return self.string(),
        else => {
            return self.errorToken("Unexpected character.");
        },
    }
}

fn isAtEnd(self: *Scanner) bool {
    return self.current >= self.source.len;
}

fn advance(self: *Scanner) void {
    self.current += 1;
}

fn advanceWithValue(self: *Scanner) u8 {
    self.advance();
    return self.source[self.current - 1];
}

fn peek(self: *Scanner) u8 {
    return self.source[self.current];
}

fn peekNext(self: *Scanner) ?u8 {
    if (self.isAtEnd()) {
        return null;
    }
    return self.source[self.current + 1];
}

fn match(self: *Scanner, expected: u8) bool {
    if (self.isAtEnd()) {
        return false;
    }
    if (self.source[self.current] != expected) {
        return false;
    }
    self.current += 1;
    return true;
}

fn makeToken(self: *Scanner, token_type: TokenType) Token {
    return .{
        .token_type = token_type,
        .start = self.source.ptr[self.start..],
        .length = self.current - self.start,
        .line = self.line,
    };
}

fn errorToken(self: *Scanner, msg: []const u8) Token {
    return .{
        .token_type = .ERROR,
        .start = msg.ptr,
        .length = msg.len,
        .line = self.line,
    };
}

fn skipWhitespace(self: *Scanner) void {
    while (true) {
        const c = self.peek();
        switch (c) {
            ' ', '\t', '\r' => self.advance(),
            '\n' => {
                self.line += 1;
                self.advance();
            },
            '/' => {
                if (self.peekNext() == '/') {
                    while (self.peek() != '\n' and !self.isAtEnd()) {
                        self.advance();
                    }
                } else {
                    return;
                }
            },
            else => return,
        }
    }
}

fn string(self: *Scanner) Token {
    while (self.peek() != '"' and !self.isAtEnd()) {
        if (self.peek() == '\n') {
            self.line += 1;
        }
        self.advance();
    }

    if (self.isAtEnd()) {
        return self.errorToken("Unterminated string.");
    }

    self.advance();
    return self.makeToken(.STRING);
}

fn number(self: *Scanner) Token {
    while (isDigit(self.peek())) {
        self.advance();
    }

    if (self.peek() == '.' and self.peekNext() != null and isDigit(self.peekNext().?)) {
        self.advance();

        while (isDigit(self.peek())) {
            self.advance();
        }
    }

    return self.makeToken(.NUMBER);
}

fn identifier(self: *Scanner) Token {
    while (isAlpha(self.peek()) or isDigit(self.peek())) {
        self.advance();
    }

    return self.makeToken(self.identifierType());
}

fn identifierType(self: *Scanner) TokenType {
    _ = self;
    return .IDENTIFIER;
}

fn isDigit(c: u8) bool {
    return std.ascii.isDigit(c);
}

fn isAlpha(c: u8) bool {
    return std.ascii.isAlphabetic(c) or c == '_';
}

const Token = struct {
    token_type: TokenType,
    start: [*]const u8,
    length: usize,
    line: usize,
};

const TokenType = enum {
    // Single-character tokens.
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    COMMA,
    DOT,
    MINUS,
    PLUS,
    SEMICOLON,
    SLASH,
    STAR,
    // One or two character tokens.
    BANG,
    BANG_EQUAL,
    EQUAL,
    EQUAL_EQUAL,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,
    // Literals.
    IDENTIFIER,
    STRING,
    NUMBER,
    // Keywords.
    AND,
    CLASS,
    ELSE,
    FALSE,
    FOR,
    FUN,
    IF,
    NIL,
    OR,
    PRINT,
    RETURN,
    SUPER,
    THIS,
    TRUE,
    VAR,
    WHILE,

    ERROR,
    EOF,
};
